src/algorithms/calorimetry/HEXPLIT.cc

```diff
@@ -0,0 +1,145 @@
+// Copyright (C) 2023 Sebouh Paul
+// SPDX-License-Identifier: LGPL-3.0-or-later
+
+// References:
+//   https://arxiv.org/abs/2308.06939
+
+
+#include <DD4hep/Readout.h>
+#include <Evaluator/DD4hepUnits.h>
+#include <TInterpreter.h>
+#include <TInterpreterValue.h>
+#include <edm4hep/Vector3d.h>
+#include <Eigen/Dense>
+#include <fmt/format.h>
+#include <map>
+#include <set>
+#include <sstream>
+#include <stdexcept>
+#include <string>
+#include <tuple>
+#include <utility>
+#include <vector>
+#include <iostream>
+
+#include "HEXPLIT.h"
+
+#include "algorithms/calorimetry/HEXPLITConfig.h"
+
+//using namespace edm4eic;
+
+namespace eicrecon {
+
+void HEXPLIT::init(const dd4hep::Detector* detector, std::shared_ptr<spdlog::logger>& logger) {
+    m_log = logger;
+    m_detector = detector;
+}
+
+std::unique_ptr<edm4eic::CalorimeterHitCollection> HEXPLIT::process(const edm4eic::CalorimeterHitCollection &hits){
+  int nhits=hits.size();
+  double sl=m_cfg.side_length/dd4hep::mm;
+  double layer_spacing=m_cfg.layer_spacing/dd4hep::mm;
+  double MIP=m_cfg.MIP/dd4hep::GeV;
+  double Emin=m_cfg.Emin_in_MIPs*MIP/dd4hep::GeV;
+  double tmax=m_cfg.tmax/dd4hep::ns;
+
+
+  double x[nhits];
+  double y[nhits];
+  double z[nhits];
+  double E[nhits];
+  double t[nhits];
+
+  for(int i=0; i<nhits; i++){
+    x[i]=hits[i].getLocal().x;
+    y[i]=hits[i].getLocal().y;
+    z[i]=hits[i].getLocal().z;
+    E[i]=hits[i].getEnergy();
+    t[i]=hits[i].getTime();
+  }
+
+  auto subcellHits = std::make_unique<edm4eic::CalorimeterHitCollection>();
+  double Esum=0;
+  for(int i=0; i<nhits; i++){
+    //skip hits that do not pass E and t cuts
+    if (E[i]<Emin || t[i]>tmax)
+      continue;
+    //keep track of the energy in each neighboring cell
+    double Eneighbors[SUBCELLS];
+    for (int j=0; j<SUBCELLS; j++)
+      Eneighbors[j]=0;
+
+    for (int j=0; j<nhits; j++){
+      //only look at hits nearby within two layers of the current layer
+      if (abs(z[i]-z[j])>2.5*layer_spacing || z[i]==z[j])
```

Comment:
Since `z` is of type double, `z[i] == z[j]` is not necessarily well defined without using epsilon.